### PR TITLE
Refactor stdlib into a separate noir library

### LINKED ISSUE #423: Refactor stdlib into a separate noir library
(From discussion with Koby)

# Problem

Currently the stdlib is treated as a rust crate, so that we can use build.rs to copy it into a configuration directory for noir. 

This is not optimal since we ideally want to view the stdlib as a standalone Noir crate that does not depend on Rust. 

# Solution

- Move the stdlib to the package manager level. When downloading nargo, it will come with a noir stdlib directory, and upon every project creation, nargo will add it as an implicit dependency.

(Implicit meaning that it should not be added to the Nargo.toml file, but via the noirc_driver API) 

# Alternatives considered

N/A

# Additional context

N/A

### PR DESCRIPTION
**Pull request by ilitteri**
# Related issue(s)

Resolves #423 

# Description

## Summary of changes

This PR abstracts the driver from the stdlib addition and moves the stdlib as a standalone nargo crate to the package manager level and when nargo is installed it is copied to the native config dir for the driver to look up to when adding it as an implicit dependency when crating a new project.

## Dependency additions / changes

`dirs` is added as a `build-dependency` in `nargo`.

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

# Additional context

N/A

### CODE REVIEW

File: crates/nargo/src/cli/mod.rs [hunk 1]
```diff
@@ -139,3 +141,59 @@ pub fn prove_and_verify(proof_name: &str, prg_dir: &Path, show_ssa: bool) -> boo
...
+    let std_crate = driver.create_non_local_crate(path_to_std_lib_file, CrateType::Library);
+    let std_crate_name = "std";
+    driver.propagate_dep(std_crate, &CrateName::new(std_crate_name).unwrap());
+}
+
+fn path_to_stdlib() -> PathBuf {
+    dirs::config_dir().unwrap().join("noir-lang").join("std/src")
+}
+
+// FIXME: I not sure that this is the right place for this tests.
```
Comment (kevaundray) [inline crates/nargo/src/cli/mod.rs:156]: Can you open an issue to track this TODO?

File: crates/noirc_driver/src/lib.rs [hunk 2]
```diff
@@ -104,10 +101,26 @@ impl Driver {
...
+
+        for crate_id in crate_ids {
+            self.context
+                .crate_graph
+                .add_dep(crate_id, dep_to_propagate_name.clone(), dep_to_propagate)
+                .expect("ice: cyclic error triggered with std library");
+        }
+    }
+
+    // NOTE: Maybe build could be skipped given that now it is a pass through method.
```
Comment (kevaundray) [inline crates/noirc_driver/src/lib.rs:121]: Can you open an issue for this to track discussion?

**Review by kevaundray — CHANGES_REQUESTED** [review]
LGTM. Just a few comments on creating tracking issues for some TODOs

**Review by guipublic — COMMENTED** [review]
I don't understand the purpose of this PR, and in particular having the std_lib not been a standalone crate anymore does not seem a good move to me.

**Comment by guipublic** [comment]
> I don't understand the purpose of this PR, and in particular having the std_lib not been a standalone crate anymore does not seem a good move to me.

We had an internal discussion and what I got from it is that the purpose of the PR is to removing dependency on the rust compiler for the std::lib, so that we do not need additional installation steps for instance.
However we agreed not to move it to a Nargo subdirectory, keep it on top-level but remove it from the compiler directory and have it managed by Nargo like you did.

**Comment by ilitteri** [comment]
> > I don't understand the purpose of this PR, and in particular having the std_lib not been a standalone crate anymore does not seem a good move to me.
> 
> We had an internal discussion and what I got from it is that the purpose of the PR is to removing dependency on the rust compiler for the std::lib, so that we do not need additional installation steps for instance. However we agreed not to move it to a Nargo subdirectory, keep it on top-level but remove it from the compiler directory and have it managed by Nargo like you did.

Great! Let me know if more changes are needed, I'll be glad to make them.

**Review by kevaundray — APPROVED** [review]

**Pull request merged**